src/userdev/java/net/minecraftforge/gradle/userdev/util/NonTaskJavaCompiler.java

```diff
@@ -0,0 +1,188 @@
+
+    private static final class Reporter implements DiagnosticListener<JavaFileObject> {
+
+        private final Logger logger;
+
+        private Reporter(Logger logger) {
+            this.logger = logger;
+        }
+
+        @Override
+        public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
+            String message = diagnostic.getMessage(null);
+            if (message.startsWith("bootstrap class path not set in conjunction with")
+                || message.startsWith("Some input files use")
+                || message.startsWith("Recompile with -Xlint:")) {
+                // use debug for these messages, it's not relevant for MC compilation
+                logger.debug(message);
+                return;
+            }
+            switch (diagnostic.getKind()) {
+                case ERROR: logger.error(message); break;
+                case WARNING:
+                case MANDATORY_WARNING: logger.warn(message); break;
+                case NOTE:
+                case OTHER: logger.lifecycle(message); break;
+            }
+        }
+    }
+
+    private static final PatternFilterable JAVA_FILES = new PatternSet();
+
+    static {
+        JAVA_FILES.include("**/*.java");
+    }
+
+    private final Logger logger;
+    private final Reporter reporter;
+    private FileCollection classpath;
+    private JavaVersion sourceCompatibility = JavaVersion.current();
+    private JavaVersion targetCompatibility = JavaVersion.current();
+    private File destinationDir;
+    private FileTree source;
+
+    public NonTaskJavaCompiler(Logger logger) {
+        this.logger = logger;
+        this.reporter = new Reporter(logger);
+    }
+
+    public void compile() {
+        JavaCompiler compiler = getJavaCompilerSafe();
+        StandardJavaFileManager fileManager = compiler.getStandardFileManager(reporter, null, StandardCharsets.UTF_8);
+
+        List<String> options = createOptions();
+        Set<File> sourceFiles = requireNonNull(source).matching(JAVA_FILES).getFiles();
+        Iterable<? extends JavaFileObject> units = createUnits(fileManager, sourceFiles);
+
+        JavaCompiler.CompilationTask task = compiler.getTask(
+            null, fileManager, reporter, options, null, units
+        );
+        logger.info("Starting compilation, compiler={}, sourceFiles={}, options={}",
+            compiler.getClass().getName(), options, sourceFiles);
+        if (!task.call()) {
+            throw new RuntimeException("Compilation failed, see logs for details.");
+        }
+    }
+
+    private JavaCompiler getJavaCompilerSafe() {
+        try {
+            return new JavaHomeBasedJavaCompilerFactory().create();
+        } catch (Exception e) {
+            // Perhaps Gradle internals were refactored. Use a simpler but maybe incorrect method.
+            logger.info("Unable to get Java compiler via Gradle, using ToolProvider", e);
+            return ToolProvider.getSystemJavaCompiler();
+        }
```

Comment:
The problem with this is javax.tools isn't a standard package that is included in every JDK. We would need to figure out a way to make sure it exists.
As for JavaHomeBasedJavaCompilerFactory, that is internal gradle API, and thus we are trying our best to stay away from it.

Reply:
I'm pretty sure that Gradle itself relies on the existence of `javax.tools`, so it would probably not be good for someone to try and build without it present anyways. I've refactored to add a command fallback anyways, since that's the best we can go for if there is no `javax.tools`.

I've also re-implemented Gradle's internal JAVA_HOME compiler resolution, so we can avoid that API as well. Although, I'm not sure if including all this extra resolution code is worth not touching the rarely-modified `Jvm` class instead.